types/merkle.proto

```diff
@@ -0,0 +1,79 @@
+syntax = "proto3";
```

Comment:
I think we want to just use go-amino structs instead of maintaining separate Merkle proto3 structures.  Also check out tendermint/iavl proofs.. I think we need to merge those proofs (e.g. range proof, absence proof) with tendermint/tmlibs/merkle simple proofs, using Go structs and interfaces, and no protobuf3.

Reply:
The other proof types must be supported. I will include them.

How about the other environments such as lotion? If we use go-amino for IBC then it will take some time to make the javascript developers use IBC. If the tradeoff is not that expensive, I will switch to go-amino.

Reply:
They have their own proof system AFAIK.  We'll work on it... I think it wouldn't be hard to create a schema language a JS library for it.

Reply:
So that helps a lot.  But right, it doesn't help all the way.

Reply:
Maybe worth keeping proto3 for IBC initially to facilitate easier implementation by other chains which can support the IBC spec but don't have Amino libraries yet? Seems like those might take awhile (or others might be hesitant to adopt a different serialization format)

Reply:
It'll be awesome to have a format that supports both SDK and JS proofs. Although it will be tough to build a whole `js-amino` library, but it seems now we need it for LCD.js and the Tendermint light client.

Also, right now I'm thinking it might be hard to make this format generic without being overly complex, so IMO it might make sense to first just write a golang verifier for JS proofs, and vice versa.

Reply:
@mappum Why do you think it would be hard to make the format generic? As-is, this protobuf spec covers most (all?) of what the SDK needs; does JS need other information?

Reply:
In trading code simplicity for efficiency, the JS tree works a lot differently. Namely, the inner nodes all have keys and values which means each inner node in the proof has a `kvHash` that must be used to calculate the node's hash along with the children hashes.

If this is already possible to implement with the current format proposal then I take back my comments, but it's not currently obvious to me.

Reply:
How about versioned encoding? For example, there could be a prefix byte for serialized Merkle proof, which indicates the encoding scheme.